base/generator.jl

```diff
@@ -52,6 +52,9 @@ size(g::Generator) = size(g.iter)
 axes(g::Generator) = axes(g.iter)
 ndims(g::Generator) = ndims(g.iter)
 
+getindex(g::Generator, I...) = map(g.f, g.iter[I...])
+firstindex(g::Generator) = firstindex(g.iter)
+lastindex(g::Generator) = lastindex(g.iter)
 
```

Comment:
How about also defining `eachindex`, `keys`, `pairs`, `IndexStyle`, etc.?

Reply:
I was trying to stick to just implementing the indexing interface but it should be fine to extend into the abstract array interface. Note only `keys` is needed for `eachindex` and `pairs`.

Reply:
`eachindex` can return something different from  (more efficient than) `keys`, so I think both are needed.

Reply:
I'm going to limit the scope of this PR just to the [indexing interface](https://docs.julialang.org/en/v1/manual/interfaces/#Indexing) as I suspect scope creep will result in this PR stalling out.